crates/ra_hir/src/source_analyzer.rs

```diff
@@ -78,9 +78,15 @@ impl SourceAnalyzer {
         }
     }
 
-    fn expr_id(&self, expr: &ast::Expr) -> Option<ExprId> {
-        let src = InFile { file_id: self.file_id, value: expr };
-        self.body_source_map.as_ref()?.node_expr(src)
+    fn expr_id(&self, db: &dyn HirDatabase, expr: &ast::Expr) -> Option<ExprId> {
```

Comment:
@matklad, what is the best way to go from `ExprId` -> `ast::Expr`. Is it `db.body_with_source_map(body_def)` query?

Reply:
Yes. In general, sources for things with ids are stored separatelly, and typically the separate storage is called the source map.

Reply:
@matklad Did you mean not ids but `Idx<T>` ?